src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -254,20 +254,21 @@ private async Task ProcessQueueAsync()
                                 throw new InvalidOperationException($"unexpected null value for {nameof(currentWorkCts)}");
                             }
 
-                            if (!concurrentlyExecutingTasks.TryAdd(currentWorkTask, currentWorkCts))
-                            {
-                                throw new InvalidOperationException($"unable to add {nameof(currentWorkTask)} into {nameof(concurrentlyExecutingTasks)}");
-                            }
-
-                            _ = currentWorkTask.ContinueWith(t =>
+                            // Only mark the task complete when it has removed itself from the executing task list.
+                            currentWorkTask = currentWorkTask.ContinueWith(t =>
                             {
-                                if (!concurrentlyExecutingTasks.TryRemove(t, out var concurrentlyExecutingTaskCts))
+                                if (!concurrentlyExecutingTasks.TryRemove(currentWorkTask, out var concurrentlyExecutingTaskCts))
```

Comment:
📝 This now has a different race condition which occurs if the continuation executes before the assignment back to the local variable

Reply:
➡️ Resolved in a different way